# To inform the algorithm to exit automatically when no data is received
autoexit:           0

# Declare the lidar topics to merge
lidar_topic:        [
                      "/os_cloud_node/points"
                    ]

merged_lidar_topic: "/merged_pc"

# Lidar extrinsics: Transform of coordinates in lidar's frame to body
# NOTE TO USER: The (4,4) entry of the extrinsics is used to declare
# whether stamp is at beginning (1.000) or end (0.000) of scan
lidar_extr:         [
                      # Velodyne
                      1, 0, 0, 0,
                      0, 1, 0, 0,
                      0, 0, 1, 0,
                      0, 0, 0, 1
                    ]

# Minimum lidar range to admit
min_range:          0.5

# IMU topic
imu_topic:          "/imu_correct"

# Gravity constant
GRAV:               9.81007

# IMU noises
GYR_N:              [ 1.5e-0 ]
GYR_W:              [ 3.5e-6 ]
ACC_N:              [ 4.0e-0 ]
ACC_W:              [ 6.4e-5 ]

# Weightage of a basic lidar factor
lidar_weight:       1.0

# Number of packets to process at time
WINDOW_SIZE:        4
# Number of substates to give to each packet
N_SUB_SEG:          8

# Downsampling grid size, chose -1 for no downsampling
leaf_size:          0.05
assoc_spacing:      0.2
surfel_map_depth:   16
surfel_min_point:   8
surfel_min_depth:   1
surfel_query_depth: 6
surfel_intsect_rad: 0.2
surfel_min_plnrty:  0.6

# Distance to surfel and min score thresholds
dis_to_surfel_max:  0.5
score_min:          0.0

# Downsample rate
ds_rate:            [
                      4
                    ]

# Keyframe evaluation
kf_min_dis:         0.5
kf_min_angle:       10

# Optimization parameters
lidar_loss_thres:   10.0

# Solver:
# Trust region strategy: "lm", "dogleg"
trustRegType:       "lm"
# Linear Algebra library used: "eigen", "lapack", "cuda"
linAlgbLib:         "cuda" 
max_solve_time:     0.3
max_iterations:     30

# Sensors fused
fuse_lidar:         1
fuse_imu:           1
fuse_poseprop:      0

regularize_imu:     1
imu_init_time:      0.1
max_outer_iters:    3
max_lidar_factor:   4000  # Maximum number of lidar factors

# Loop closure parameters
loop_en:            1    # Enable loop closure
loop_kf_nbr:        10   # Number of neighbour to check for loop closure
loop_time_mindiff:  10   # Only check for loop when keyframes have this much difference

icpMaxIter:         100   # Maximum iterations for ICP
icpFitnessThres:    0.3   # Fitness threshold for ICP check
histDis:            15    # Maximum correspondence distance for icp

# Bundle adjustment params
rib_edge:           5
odom_q_noise:       0.1
odom_p_noise:       0.1
loop_weight:        0.1

# Number of optimizations before quitting
debug_exit:        -1
publish_map:        0

# Log directory
log_dir:            "/home/tmn/slict_log/ntu_viral"